### PR TITLE
Cor 124 cms api validation

### PR DESCRIPTION
**Pull request by nilueps**
So this probably needs some work, what do you guys think?

### CODE REVIEW

**Review by shinroo — COMMENTED** [review]
Looks good to me, like the popover and changes to the response messages!

**Comment by nilueps** [comment]
@ludydoo can you check this?

**Review by ludydoo — APPROVED** [review]

File: api/pkg/apps/cms/case_post.go [hunk 1]
```diff
@@ -15,6 +16,19 @@ func (s *Server) PostCase(w http.ResponseWriter, req *http.Request) {
...
+
+	errList := ValidateCase(kase, &validation.Path{})
+	if len(errList) > 0 {
+		status := validation.Status{
+			Status:  validation.Failure,
+			Code:    http.StatusUnprocessableEntity,
+			Message: "invalid case type",
+			Errors:  errList,
+		}
+		s.JSON(w, status.Code, status)
```
Comment (ludydoo) [inline api/pkg/apps/cms/case_post.go:28]: Same there here to extract the method

File: api/pkg/apps/cms/case_validation.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+func ValidateCase(kase *Case, path *validation.Path) validation.ErrorList {
+	errList := validation.ErrorList{}
+
+	// Validate form elements
+	for _, elem := range kase.FormData.FormElements {
+		if elem.Validation.Required && len(elem.Attributes.Value) == 0 {
+			errList = append(errList, validation.Required(path.Child(elem.Attributes.ID), fmt.Sprintf("%s is required", elem.Attributes.ID)))
+		}
+		// TODO implement validation for specific input controls (email, date, etc)
```
Comment (ludydoo) [inline api/pkg/apps/cms/case_validation.go:16]: Jira issue ?

File: api/pkg/apps/cms/casetype_validation.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+func ValidateCaseType(caseType *CaseType, path *validation.Path) validation.ErrorList {
+	errList := validation.ErrorList{}
+
+	if len(caseType.Name) == 0 {
+		errList = append(errList, validation.Required(path.Child("name"), "name is required"))
+	} else if !validation.IsValidAlpha(caseType.Name) {
+		errList = append(errList, validation.Invalid(path.Child("name"), caseType.Name, "name should only contain letters and spaces"))
+	}
+	if len(caseType.PartyTypeID) == 0 {
```
Comment (ludydoo) [inline api/pkg/apps/cms/casetype_validation.go:13]: We should also make sure that all UUIDS are actually valid UUIDs
```
func IsValidUUID(str string) bool {
_, err := uuid.FromString(str);
return err == nil
}
```

File: api/pkg/apps/webapp/casetypes.go [hunk 4]
```diff
@@ -17,89 +17,29 @@ func (s *Server) CaseTypes(w http.ResponseWriter, req *http.Request) {
...
-		s.PostCaseType(ctx, &cms.CaseType{}, w, req)
-		return
-	}
-
 	caseTypes, err := cmsClient.CaseTypes().List(ctx, cms.CaseTypeListOptions{})
-	if err != nil {
-		s.Error(w, err)
+	if errResponse(w, err) {
 		return
 	}
```
Comment (ludydoo) [inline api/pkg/apps/webapp/casetypes.go:23]: You'll probably have merge conflicts on master for these

File: api/pkg/apps/webapp/casetypes.go [hunk 5]
```diff
@@ -132,60 +70,120 @@ func (s *Server) CaseType(w http.ResponseWriter, req *http.Request) {
...
+	req *http.Request,
+	partyTypes *iam.PartyTypeList,
+	teamsData *iam.TeamList,
+) {
+	cmsClient := h.CMSClient(ctx)
+
+	isNew := false
+	if len(caseType.ID) == 0 {
+		isNew = true
+	}
```
Comment (ludydoo) [inline api/pkg/apps/webapp/casetypes.go:115]: minor
```
isNew := len(caseType.ID) == 0
```

File: api/pkg/apps/webapp/casetypes.go [hunk 6]
```diff
@@ -132,60 +70,120 @@ func (s *Server) CaseType(w http.ResponseWriter, req *http.Request) {
...
+
+	if err := req.ParseForm(); errResponse(w, err) {
+		return
+	}
```
Comment (ludydoo) [inline api/pkg/apps/webapp/casetypes.go:119]: I think the errResponse thing is a bit hard to read..

**Review by ludydoo — CHANGES_REQUESTED** [review]

File: api/pkg/apps/cms/casetype_validation.go (hunk 3 above)
Comment (nilueps) [reply api/pkg/apps/cms/casetype_validation.go:13]: I assume that means adding a subroutine in the Status method to check for uuid errors and return a different error type in that case? The front-end user can't correct that type of error, right?

Is this really necessary though? All UUIDs are generated by the same library

File: api/pkg/apps/cms/case_post.go [hunk 7]
```diff
@@ -15,6 +16,14 @@ func (s *Server) PostCase(w http.ResponseWriter, req *http.Request) {
 	}
 
 	kase := &payload
+
+	errList := ValidateCase(kase, &validation.Path{})
+	if len(errList) > 0 {
+		status := errList.Status(http.StatusUnprocessableEntity, "invalid case")
+		s.JSON(w, status.Code, status)
+		return
```
Comment (ludydoo) [inline api/pkg/apps/cms/case_post.go:24]: s.Error(w, status)

File: api/pkg/apps/cms/casetype_validation.go (hunk 3 above)
Comment (ludydoo) [reply api/pkg/apps/cms/casetype_validation.go:13]:
```
func ValidateUUID(uuid string, path validation.Path) validation.ErrList {...}
```

**Review by ludydoo — APPROVED** [review]

**Pull request merged**